File: pdv-python/pdv/serializers.py
"""
pdv.serializers — Custom serializer registry and dunder protocol for custom types.

PDV supports two extension paths for persisting instances of a custom class:

1. **Registered serializers.** A module developer calls
   ``pdv.register_serializer(MyClass, format=..., extension=..., save=..., load=...)``
   to attach save/load callbacks to a class they may not own (e.g.
   ``scipy.sparse.csr_matrix``). This is the only supported way to save
   objects whose state lives outside Python (ctypes pointers, Fortran library
   handles, GPU buffers, ...).
2. **Dunder protocol.** A class the package author defines may opt in by
   implementing ``__pdv_format__`` / ``__pdv_serialize__`` /
   ``__pdv_deserialize__`` (required as a set) plus the optional
   ``__pdv_preview__``, ``__pdv_handle__``, and ``__pdv_digest__`` methods.
   The defining package never imports ``pdv``; load-time class recovery uses
   the descriptor's ``python_type`` metadata and ``importlib.import_module``.

Registered serializers take precedence over the dunder protocol when both
exist for the same class — explicit registration is the way to override a
class's own intent for types you don't own.

Public API
----------
register : function
    ``pdv.register_serializer(MyClass, format=..., extension=..., save=..., load=...)``.
find_for_value : function
    Look up a registered entry by walking ``type(value).__mro__``.
find_for_format : function
    Look up a registered entry by its format name (used during load).
find_for_value_dunder : function
    Synthesize a :class:`DunderEntry` from a value's dunder methods, or None.
find_for_format_dunder : function
    Recover the class for a dunder-served format by importing ``python_type``.
get_registry : function
    Snapshot of registered serializers, for tests and debugging.
clear : function
    Drop all registered serializers and dunder caches (used in tests).

See Also
--------
pdv.serialization (consumer of this registry)
pdv.modules (companion ``@pdv.handle`` decorator and ``__pdv_handle__`` fallback)
ARCHITECTURE.md §7.2 (node types)
"""

from __future__ import annotations

import importlib
from dataclasses import dataclass
from typing import Any, Callable, Optional

from pdv.errors import PDVSerializationError


@dataclass
class SerializerEntry:
    """One registered serializer mapping a type to save/load callbacks."""

    cls: type
    format: str
    extension: str
    save: Callable[[Any, str], None]
    load: Callable[[str], Any]
    preview: Optional[Callable[[Any], str]]
    class_name: str  # fully qualified: "module.Class"


_serializer_registry: dict[type, SerializerEntry] = {}
_format_index: dict[str, SerializerEntry] = {}


@dataclass
class DunderEntry:
    """A class's dunder-protocol opt-in, synthesized from its methods."""

    cls: type
    format: str
    extension: str
    class_name: str  # fully qualified: "module.Class"


# Per-type cache; None marks a class that has been checked and found not to
# implement the full trio. Cleared by :func:`clear`.
_dunder_cache: dict[type, Optional[DunderEntry]] = {}

# Maps (format, python_type) -> class object so we don't re-import on every
# subsequent load of the same format.
_dunder_format_cache: dict[tuple[str, str], type] = {}


# Format names reserved by builtin serializers in pdv.serialization.
# Listed inline (rather than imported) to avoid a circular import — the
# constants in serialization.py are the source of truth and these strings
# must stay in sync.
_RESERVED_FORMATS: frozenset[str] = frozenset(
    {
        "npy",
        "json",
        "txt",
        "pickle",
        "py_script",
        "markdown",
        "inline",
        "gui_json",
        "module_meta",
        "namelist",
        "py_lib",
        "bin",
        "none",
    }
)


def register(
    cls: type,
    *,
    format: str,
    extension: str = ".bin",
    save: Callable[[Any, str], None],
    load: Callable[[str], Any],
    preview: Optional[Callable[[Any], str]] = None,
) -> None:
    """Register a custom serializer for instances of *cls*.

    PDV will choose the on-disk filename and pass an absolute path to the
    ``save`` callback. The callback writes the object's state to that path
    however it likes. ``load`` receives the same path on project load and
    must return a reconstructed instance.

    Parameters
    ----------
    cls : type
        The type to register a serializer for. Lookup walks the MRO, so a
        serializer registered on a base class also covers subclasses.
    format : str
        Unique format identifier stored in ``tree-index.json`` so the right
        loader can be found at load time. Must not collide with any builtin
        format name (``npy``, ``pickle``, ``json``, ...).
    extension : str
        File extension PDV appends to the chosen filename. Leading ``.`` is
        added if missing. Defaults to ``".bin"``.
    save : callable
        ``save(obj, abs_path) -> None``. Writes the object's state to
        ``abs_path``. May raise; PDV wraps errors with the tree path.
    load : callable
        ``load(abs_path) -> obj``. Reads the file PDV wrote and returns the
        reconstructed instance.
    preview : callable, optional
        ``preview(obj) -> str``. Short human-readable preview shown in the
        tree panel. Falls back to a generic ``"<ClassName>"`` string when
        not supplied.

    Raises
    ------
    PDVSerializationError
        If ``format`` is empty, collides with a builtin format, or
        ``cls`` is not a class.
    """
    if not isinstance(cls, type):
        raise PDVSerializationError(
            f"register_serializer: first argument must be a class, got {type(cls).__name__}"
        )
    if not isinstance(format, str) or not format:
        raise PDVSerializationError(
            "register_serializer: 'format' must be a non-empty string"
        )
    if format in _RESERVED_FORMATS:
        raise PDVSerializationError(
            f"register_serializer: format '{format}' collides with a builtin format name"
        )
    if not callable(save) or not callable(load):
        raise PDVSerializationError(
            "register_serializer: 'save' and 'load' must be callables"
        )

    ext = extension or ".bin"
    if not ext.startswith("."):
        ext = "." + ext

    class_name = f"{cls.__module__}.{cls.__qualname__}"

    if format in _format_index and _format_index[format].cls is not cls:
        import warnings  # noqa: PLC0415

        old = _format_index[format]
        warnings.warn(
            f"Serializer format '{format}' overwritten "
            f"(was {old.class_name}, now {class_name})"
        )
        # Drop the old class binding so the registry stays consistent.
        _serializer_registry.pop(old.cls, None)

    if cls in _serializer_registry:
        import warnings  # noqa: PLC0415

        old = _serializer_registry[cls]
        warnings.warn(
            f"Serializer for {class_name} overwritten (was format '{old.format}')"
        )
        _format_index.pop(old.format, None)

    entry = SerializerEntry(
        cls=cls,
        format=format,
        extension=ext,
        save=save,
        load=load,
        preview=preview,
        class_name=class_name,
    )
    _serializer_registry[cls] = entry
    _format_index[format] = entry


def find_for_value(value: Any) -> Optional[SerializerEntry]:
    """Return the registered serializer matching *value*'s type, or None.

    Walks ``type(value).__mro__`` so subclasses inherit a base class's
    registration.
    """
    for cls in type(value).__mro__:
        entry = _serializer_registry.get(cls)
        if entry is not None:
            return entry
    return None


def find_for_format(format: str) -> Optional[SerializerEntry]:
    """Return the registered serializer for *format*, or None."""
    return _format_index.get(format)


def get_registry() -> dict[str, str]:
    """Return ``{class_name: format}`` for all registered serializers."""
    return {entry.class_name: entry.format for entry in _serializer_registry.values()}


def find_for_value_dunder(value: Any) -> Optional[DunderEntry]:
    """Return a :class:`DunderEntry` for *value*'s class, or ``None``.

    A class opts into the dunder protocol by defining all three of
    ``__pdv_format__`` (classmethod returning ``(format_name, extension)``),
    ``__pdv_serialize__`` (instance method writing the value to a path), and
    ``__pdv_deserialize__`` (classmethod reading the file back). When a class
    defines only one or two of those (the dunder names may be present on the
    class for unrelated reasons), this function returns ``None`` so the
    caller can fall through to the pickle fallback.

    Lookup is via standard ``hasattr`` on ``type(value)``, which walks the
    MRO. The result is cached per type.

    Parameters
    ----------
    value : Any
        The value PDV is about to serialize.

    Returns
    -------
    DunderEntry or None
        ``None`` when the class does not implement the full trio.

    Raises
    ------
    PDVSerializationError
        If ``__pdv_format__()`` raises, returns a non-(str, str) tuple, or
        returns a name that collides with a builtin format or with a
        :func:`register` entry for a *different* class.
    """
    cls = type(value)
    cached = _dunder_cache.get(cls)
    if cached is not None:
        return cached
    if cls in _dunder_cache:  # cached as None — definitively no protocol
        return None

    has_fmt = hasattr(cls, "__pdv_format__")
    has_ser = hasattr(cls, "__pdv_serialize__")
    has_des = hasattr(cls, "__pdv_deserialize__")
    present = sum((has_fmt, has_ser, has_des))
    if present < 3:
        # Partial trio: silently treat as "no protocol" so users may use any
        # of these names for unrelated reasons. Cache the negative result so
        # repeated saves are cheap.
        _dunder_cache[cls] = None
        return None

    class_name = f"{cls.__module__}.{cls.__qualname__}"

    try:
        result = cls.__pdv_format__()
    except Exception as exc:  # noqa: BLE001
        raise PDVSerializationError(
            f"Dunder protocol for '{class_name}': __pdv_format__() raised: {exc}"
        ) from exc

    if (
        not isinstance(result, tuple)
        or len(result) != 2
        or not isinstance(result[0], str)
        or not isinstance(result[1], str)
    ):
        raise PDVSerializationError(
            f"Class '{class_name}.__pdv_format__()' must return a 2-tuple of "
            f"(format_name: str, extension: str). Got: {result!r}"
        )

    fmt, ext = result
    if fmt in _RESERVED_FORMATS:
        raise PDVSerializationError(
            f"Dunder protocol for '{class_name}': format '{fmt}' collides with "
            f"a builtin format name. Choose a different name in __pdv_format__()."
        )

    existing = _format_index.get(fmt)
    if existing is not None and existing.cls is not cls:
        raise PDVSerializationError(
            f"Dunder protocol for '{class_name}': format '{fmt}' is already "
            f"registered via pdv.register_serializer for "
            f"'{existing.class_name}'. Choose a different name in "
            f"__pdv_format__() or remove the conflicting registration."
        )

    if not ext.startswith("."):
        ext = "." + ext

    entry = DunderEntry(cls=cls, format=fmt, extension=ext, class_name=class_name)
    _dunder_cache[cls] = entry
    return entry


def find_for_format_dunder(fmt: str, python_type: str) -> Optional[type]:
    """Recover a class implementing ``__pdv_deserialize__`` by importing it.

    On project load, PDV reads the descriptor's ``metadata.python_type``
    (e.g. ``"mypkg.geqdsk.GEqdskData"``) and asks this function to import the
    module, walk to the named class, and confirm it implements the dunder
    protocol. The defining package only needs to be installed — it does
    **not** need to be imported by the user before project load.

    Never raises: returns ``None`` for any failure mode (missing module,
    unknown attribute, class no longer implementing the protocol). The
    caller is expected to raise a descriptive ``PDVSerializationError`` that
    names the format and the ``python_type`` so the user knows what to install.

    Parameters
    ----------
    fmt : str
        The on-disk format identifier read from ``storage.format``. Used as
        part of the cache key.
    python_type : str
        Dotted path ``"module.qualname"``. Supports nested ``qualname`` via
        the inner-dot walk after the module split.

    Returns
    -------
    type or None
        The recovered class, or ``None`` if the import/resolution failed.
    """
    if not python_type:
        return None
    cache_key = (fmt, python_type)
    cached = _dunder_format_cache.get(cache_key)
    if cached is not None:
        return cached

    # Split on the *last* dot first; on failure walk leftwards so nested
    # qualnames like "pkg.mod.Outer.Inner" still resolve.
    head, sep, tail = python_type.rpartition(".")
    while head:
        try:
            module = importlib.import_module(head)
        except Exception:  # noqa: BLE001
            head, sep, rest = head.rpartition(".")
            tail = f"{rest}.{tail}" if sep else tail
            continue
        obj: Any = module
        try:
            for part in tail.split("."):
                obj = getattr(obj, part)
        except AttributeError:
            return None
        if not isinstance(obj, type):
            return None
        if not hasattr(obj, "__pdv_deserialize__"):
            return None
        _dunder_format_cache[cache_key] = obj
        return obj
    return None


def clear() -> None:
    """Drop all registered serializers and dunder caches. Used in tests."""
    _serializer_registry.clear()
    _format_index.clear()
    _dunder_cache.clear()
    _dunder_format_cache.clear()

File: pdv-julia/src/serializers.jl
# serializers.jl — Custom serializer registry and the PDV protocol functions.
#
# Port of pdv/serializers.py. Two extension paths for persisting instances of
# a custom type:
#
# 1. **Registered serializers.** `register_serializer(T; format, extension,
#    save, load, preview)` attaches save/load callbacks to a type the caller
#    may not own.
# 2. **Protocol methods** (Julia analog of Python's dunder protocol). A
#    package opts its own type in by adding methods to PDVKernel's generic
#    functions:
#
#        PDVKernel.pdv_format(::Type{MyType}) = ("my_fmt", ".h5")
#        PDVKernel.pdv_serialize(obj::MyType, abs_path::String) = ...
#        PDVKernel.pdv_deserialize(::Type{MyType}, abs_path::String) = MyType(...)
#
#    plus the optional `pdv_preview(obj)`, `pdv_handle(obj, path, tree)`, and
#    `pdv_digest(obj)` methods. Load-time type recovery resolves the stored
#    type string by walking module paths from `Main`.
#
# Registered serializers win over protocol methods when both exist.

# ---------------------------------------------------------------------------
# Protocol generic functions (modules add methods; PDVKernel defines none)
# ---------------------------------------------------------------------------

"""
    pdv_format(::Type{T}) -> (format_name::String, extension::String)

Protocol hook: declare the on-disk format for values of type `T`.
Required (together with [`pdv_serialize`](@ref) and [`pdv_deserialize`](@ref))
to opt a type into PDV persistence.
"""
function pdv_format end

"""
    pdv_serialize(obj, abs_path::AbstractString)

Protocol hook: write `obj`'s state to `abs_path`. PDV chooses the path.
"""
function pdv_serialize end

"""
    pdv_deserialize(::Type{T}, abs_path::AbstractString) -> T

Protocol hook: read the file written by [`pdv_serialize`](@ref) and return a
reconstructed instance.
"""
function pdv_deserialize end

"""
    pdv_preview(obj) -> String

Protocol hook: short human-readable preview shown in the tree panel.
"""
function pdv_preview end

"""
    pdv_handle(obj, path::AbstractString, tree)

Protocol hook: double-click handler for tree nodes holding values of the
method's type. The Julia analog of Python's `@pdv.handle` decorator — modules
register handlers by adding methods (see the bundled N-pendulum-julia module).
"""
function pdv_handle end

"""
    pdv_digest(obj) -> Vector{UInt8} | String

Protocol hook: stable byte payload used for change-detection checksums.
"""
function pdv_digest end

# hasmethod checks against these signatures:
_has_protocol_trio(::Type{T}) where {T} =
    hasmethod(pdv_format, Tuple{Type{T}}) &&
    hasmethod(pdv_serialize, Tuple{T,String}) &&
    hasmethod(pdv_deserialize, Tuple{Type{T},String})

# ---------------------------------------------------------------------------
# Registered serializer registry
# ---------------------------------------------------------------------------

"""One registered serializer mapping a type to save/load callbacks."""
struct SerializerEntry
    type::Type
    format::String
    extension::String
    save::Function
    load::Function
    preview::Union{Nothing,Function}
    type_name::String  # fully qualified, e.g. "Main.NPendulum.PendulumSolution"
end

const _SERIALIZER_REGISTRY = Dict{Type,SerializerEntry}()
const _FORMAT_INDEX = Dict{String,SerializerEntry}()

# Format names reserved by builtin serializers in serialization.jl. Includes
# the Python-side names too so a project can never carry an ambiguous format.
const _RESERVED_FORMATS = Set([
    "npy", "json", "txt", "pickle", "jls", "py_script", "jl_script",
    "markdown", "inline", "gui_json", "module_meta", "namelist",
    "py_lib", "jl_lib", "bin", "file", "none", "netcdf", "hdf5",
])

"""
    fully_qualified_type_name(T::Type) -> String

Return the module-qualified name of `T` (e.g. `"Main.NPendulum.PendulumSolution"`).
"""
function fully_qualified_type_name(T::Type)::String
    base = Base.unwrap_unionall(T)
    if base isa DataType
        return string(parentmodule(base)) * "." * string(nameof(base))
    end
    return string(T)
end

"""
    register_serializer(T::Type; format, extension=".bin", save, load, preview=nothing)

Register a custom serializer for instances of `T`. PDV chooses the on-disk
filename and passes an absolute path to `save(obj, abs_path)`; `load(abs_path)`
must return a reconstructed instance at project-load time. Lookup walks the
type hierarchy, so a serializer registered on an abstract type also covers
subtypes.

Throws `PDVSerializationError` when `format` is empty or collides with a
builtin format name.
"""
function register_serializer(T::Type; format::AbstractString,
                             extension::AbstractString=".bin",
                             save::Function, load::Function,
                             preview::Union{Nothing,Function}=nothing)
    isempty(format) &&
        throw(PDVSerializationError("register_serializer: 'format' must be a non-empty string"))
    format in _RESERVED_FORMATS &&
        throw(PDVSerializationError(
            "register_serializer: format '$format' collides with a builtin format name"))

    ext = isempty(extension) ? ".bin" : String(extension)
    startswith(ext, ".") || (ext = "." * ext)
    type_name = fully_qualified_type_name(T)

    fmt = String(format)
    if haskey(_FORMAT_INDEX, fmt) && _FORMAT_INDEX[fmt].type !== T
        old = _FORMAT_INDEX[fmt]
        @warn "Serializer format '$fmt' overwritten (was $(old.type_name), now $type_name)"
        delete!(_SERIALIZER_REGISTRY, old.type)
    end
    if haskey(_SERIALIZER_REGISTRY, T)
        old = _SERIALIZER_REGISTRY[T]
        @warn "Serializer for $type_name overwritten (was format '$(old.format)')"
        delete!(_FORMAT_INDEX, old.format)
    end

    entry = SerializerEntry(T, fmt, ext, save, load, preview, type_name)
    _SERIALIZER_REGISTRY[T] = entry
    _FORMAT_INDEX[fmt] = entry
    nothing
end

"""
    find_for_value(value) -> Union{SerializerEntry,Nothing}

Return the registered serializer matching `value`'s type, walking the type
hierarchy so subtypes inherit a supertype's registration.
"""
function find_for_value(value)
    T = typeof(value)
    while true
        entry = get(_SERIALIZER_REGISTRY, T, nothing)
        entry !== nothing && return entry
        T === Any && return nothing
        T = supertype(T)
    end
end

"""Return the registered serializer for `format`, or `nothing`."""
find_for_format(format::AbstractString) = get(_FORMAT_INDEX, format, nothing)

"""Return `{type_name => format}` for all registered serializers."""
get_serializer_registry() =
    Dict(entry.type_name => entry.format for entry in values(_SERIALIZER_REGISTRY))

"""Drop all registered serializers (used in tests)."""
function clear_serializers!()
    empty!(_SERIALIZER_REGISTRY)
    empty!(_FORMAT_INDEX)
    nothing
end

# ---------------------------------------------------------------------------
# Protocol lookup (value side and load side)
# ---------------------------------------------------------------------------

"""A type's protocol opt-in, synthesized from its `pdv_*` methods."""
struct ProtocolEntry
    type::Type
    format::String
    extension::String
    type_name::String
end

"""
    find_for_value_protocol(value) -> Union{ProtocolEntry,Nothing}

Return a `ProtocolEntry` when `typeof(value)` implements the full
`pdv_format` / `pdv_serialize` / `pdv_deserialize` trio, else `nothing`.

Throws `PDVSerializationError` when `pdv_format` misbehaves or its format
name collides with a builtin or a registered serializer for a different type.
"""
function find_for_value_protocol(value)
    T = typeof(value)
    _has_protocol_trio(T) || return nothing
    type_name = fully_qualified_type_name(T)

    result = try
        Base.invokelatest(pdv_format, T)
    catch err
        throw(PDVSerializationError(
            "PDV protocol for '$type_name': pdv_format() raised: $err"))
    end
    if !(result isa Tuple && length(result) == 2 &&
         result[1] isa AbstractString && result[2] isa AbstractString)
        throw(PDVSerializationError(
            "pdv_format(::Type{$type_name}) must return a 2-tuple of " *
            "(format_name::String, extension::String). Got: $(repr(result))"))
    end
    fmt, ext = String(result[1]), String(result[2])
    fmt in _RESERVED_FORMATS && throw(PDVSerializationError(
        "PDV protocol for '$type_name': format '$fmt' collides with a builtin " *
        "format name. Choose a different name in pdv_format()."))
    existing = get(_FORMAT_INDEX, fmt, nothing)
    if existing !== nothing && existing.type !== T
        throw(PDVSerializationError(
            "PDV protocol for '$type_name': format '$fmt' is already registered " *
            "via register_serializer for '$(existing.type_name)'."))
    end
    startswith(ext, ".") || (ext = "." * ext)
    return ProtocolEntry(T, fmt, ext, type_name)
end

# Failure reason codes returned by find_for_format_protocol.
const LOOKUP_OK = ""
const LOOKUP_NO_TYPE = "no_type"
const LOOKUP_IMPORT_FAILED = "import_failed"
const LOOKUP_TYPE_UNLOADABLE = "type_unloadable"

"""
    resolve_type_string(type_string) -> Union{Type,Nothing}

Resolve a dotted, module-qualified type name (e.g.
`"Main.NPendulum.PendulumSolution"`) to a live `Type`, walking `getfield`
from the named root module. Roots tried: `Main`, then any loaded package of
the first segment's name. Returns `nothing` when unresolvable.
"""
function resolve_type_string(type_string::AbstractString)
    isempty(type_string) && return nothing
    parts = split(type_string, ".")
    # Determine the root module.
    obj::Any = nothing
    rest = parts
    first_seg = Symbol(parts[1])
    if first_seg === :Main
        obj = Main
        rest = parts[2:end]
    elseif isdefined(Main, first_seg) && getfield(Main, first_seg) isa Module
        obj = getfield(Main, first_seg)
        rest = parts[2:end]
    else
        for (_, mod) in Base.loaded_modules
            if nameof(mod) === first_seg
                obj = mod
                rest = parts[2:end]
                break
            end
        end
    end
    obj === nothing && return nothing
    for part in rest
        sym = Symbol(part)
        (obj isa Module || obj isa Type) || return nothing
        isdefined(obj, sym) || return nothing
        obj = getfield(obj, sym)
    end
    return obj isa Type ? obj : nothing
end

"""
    find_for_format_protocol(fmt, type_string) -> (Union{Type,Nothing}, reason)

Recover a type implementing `pdv_deserialize` from the descriptor's stored
type string at project-load time. Never throws; returns a failure reason
code so the caller can render an actionable error.
"""
function find_for_format_protocol(fmt::AbstractString, type_string::AbstractString)
    isempty(type_string) && return (nothing, LOOKUP_NO_TYPE)
    T = try
        resolve_type_string(type_string)
    catch
        nothing
    end
    T === nothing && return (nothing, LOOKUP_IMPORT_FAILED)
    hasmethod(pdv_deserialize, Tuple{Type{T},String}) || return (nothing, LOOKUP_TYPE_UNLOADABLE)
    return (T, LOOKUP_OK)
end
